Posts/2021/02/XML-NewEmployees.md

```diff
@@ -0,0 +1,180 @@
+```yaml
+---
+post_title: 'Leveraging XML with New Employee PowerShell Scripts'
+username: s-r-turner
+categories: ActiveDirectory, XML, PowerShell
+tags: XML, ActiveDirectory
+featured_image:
+summary: using xml referencing to bolster up your new employee scripts with information.
+---
+```
+
+# Leveraging XML With New Employee PowerShell Scripts
+
+I'm going to show how you can leverage XML files for referencing information, to help bolster up your Active Directory user accounts.
+
+## Scenario
+
+You have multiple subsidiary companies that each have their own individual sets of offices and addresses. You need to ensure that each new user you create has the correct address and contact information based on where they are working.
+
+## Getting Started - Creating the XML
+
+To create the xml, we're going to use the XmlWriter class, put this into a variable to build upon and add some formatting options. Like so:
+
+```powershell
+$XMLPath = "C:\Users\sam\AD-References.xml"
+$XMLWriter = New-Object System.XML.XMlTextWriter($XMLPath,$null)
+$xmlWriter.Formatting = 'Indented'
+$xmlWriter.Indentation = 1
+$XmlWriter.IndentChar = "`t"
+```
+If you run that, nothing crazy happens. We have a new xml source file, with nothing inside; we have set ourselves up to flesh it out.
+
+```powershell
+# write the header
+$xmlWriter.WriteStartDocument()
+$XMLWriter.WriteStartElement('Companies')
+# set XSL statements
+$xmlWriter.WriteProcessingInstruction("xml-stylesheet", "type='text/xsl' href='style.xsl'")
+```
+Okay awesome, now we're getting somewhere. We have started our "Companies" element. Let's create a company and add some office information beneath that. Then we'll finalise the XML document.
+
+```powershell
+# Start Company 1
+$XMLWriter.WriteStartElement('Company-1')
+
+# Start UK Offices
+$XMLWriter.WriteStartElement('UK-Offices')
+
+# Create Birmingham Office
+$xmlWriter.WriteStartElement('Birmingham')
+$XmlWriter.WriteElementString('Street', 'Unit 77, 132 Dummy Lane')
+$XmlWriter.WriteElementString('City', 'Birmingham')
+$XMLWriter.WriteElementString('State', 'Staffordshire')
+$XMLWriter.WriteElementString('Postcode', 'B1 1BB')
+$XMLWriter.WriteElementString('Country', 'GB')
+$XMLWriter.WriteEndElement()
+
+# End UK Offices
+$XMLWriter.WriteEndElement()
+
+#End Company 1
+$XMLWriter.WriteEndElement()
+
+# Finish up document
+$XMLWriter.WriteEndElement()
+$xmlWriter.WriteEndDocument()
+$xmlWriter.Flush()
+$xmlWriter.Close()
+```
+
+So, let's see the finished XML after we run this code.
+
+```xml
+<?xml version="1.0"?>
+<Companies>
+	<?xml-stylesheet type='text/xsl' href='style.xsl'?>
+	<Company-1>
+		<UK-Offices>
+			<Birmingham>
+				<Street>Unit 77, 132 Dummy Lane</Street>
+				<City>Birmingham</City>
+				<State>Staffordshire</State>
+				<Postcode>B1 1BB</Postcode>
+				<Country>GB</Country>
+			</Birmingham>
+		</UK-Offices>
+	</Company-1>
+</Companies>
```

Comment:
I thought indentation was 1 character, this looks like more?
can you shrink down the indentation a bit - say to 2 characters

Reply:
2 will indent further. I think the issue is the number of root elements in the XML. I can cut it down for the purpose of the post.